system/gsettings.py

```diff
@@ -0,0 +1,248 @@
+            except:
+                return None
+
+        def proc_owner(pid):
+            """Return username of UID of process pid"""
+            UID = 1
+            for ln in open('/proc/{0}/status'.format(pid), 'r'):
+                if ln.startswith('Uid:'):
+                    return int(ln.split()[UID])
+
+        def proc_environ(pid, varname=""):
+            """Return process environment variable of process pid"""
+            proc_env = open("/proc/{0}/environ".format(pid), 'r')
+            env = proc_env.read()
+            for env_var in env.split('\x00'):
+                if varname != "":
+                    prefix = "{0}=".format(varname)
+                    if env_var.startswith(prefix):
+                        return env_var[len(prefix):]
+
+        def bus_address(pid):
+            """Return the bus address for the specified process"""
+            return proc_environ(pid, ENV_DBUS)
+
+        def new_session():
+            """Create a new dbus session"""
+            dbus_process = subprocess.Popen('dbus-launch', stdout=subprocess.PIPE)
+            for env_string in dbus_process.stdout.readlines():
+                env_var, env_value = env_string.split('=', 1)
+                os.environ[env_var] = env_value
+            return os.environ['DBUS_SESSION_BUS_PID'].replace('\n', '')
+
+        # Determine the session pid.
+        self.session_pid = -1
+        self._cleanup = False
+
+        for procfile in os.listdir('/proc'):
+            if procfile.isdigit():
+                if proc_name(procfile) in SESSION_MANAGERS:
+                    self.session_pid = procfile
+
+        osuser = os.getuid()
+        if self.session_pid != -1:
+            sessionuser = proc_owner(self.session_pid)
+        else:
+            sessionuser = osuser
+
+        # Setup the environment and connect to the bus.
+        if self.session_pid != -1 and osuser == sessionuser:
+            os.environ[ENV_DBUS] = bus_address(self.session_pid)
+        else:
+            self.session_pid = new_session()
+            self._cleanup = True
+
+        self.bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
+
+    def get_value(self, schema, key):
+        """ Read the value of the specified key """
+        if ':' in schema:
+            schema, path = schema.split(':', 1)
+        else:
+            path = None
+
+        # Gio.Settings does not fail gracefully, making
+        # use of abort when an exception occurs.  Thus,
+        # careful checking of values is done before
+        # commiting to get_value.
+        if schema in Gio.Settings.list_schemas():
+            settings = Gio.Settings(schema=schema, path=path)
+            if key in Gio.Settings.list_keys(settings):
+                return settings.get_value(key)
+            else:
+                raise ValueError('{0} is not valid in the schema: {1}'.format(key, schema))
+        else:
+            raise ValueError('{0} is not a valid schema'.format(schema))
```

Comment:
This would flow a bit more naturally without nesting the checks:

```python
if schema not in ...list_schemas():
    raise ...
settings = ...
if key not in ...list_keys(...):
    raise ...
return settings.get_value(key)
```